From ea7fea87a2cdba9df40f27464ce27f5d192ac389 Mon Sep 17 00:00:00 2001
From: Alexander Sulfrian <alexander.sulfrian@fu-berlin.de>
Date: Mon, 7 Jan 2013 19:11:49 +0100
Subject: [PATCH 1/2] add new keyCodes used by >=FF15

---
 shellinabox/vt100.jspp | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/shellinabox/vt100.jspp b/shellinabox/vt100.jspp
index 7a72d0c..578a1b6 100755
--- a/shellinabox/vt100.jspp
+++ b/shellinabox/vt100.jspp
@@ -2894,10 +2894,10 @@ VT100.prototype.keyDown = function(event) {
     event.keyCode == 226;
   var normalKey                 =
     alphNumKey                                   ||
-    event.keyCode ==  59 || event.keyCode ==  61 ||
+    event.keyCode >=  59 && event.keyCode <=  64 ||
     event.keyCode == 106 || event.keyCode == 107 ||
     event.keyCode >= 109 && event.keyCode <= 111 ||
-    event.keyCode >= 186 && event.keyCode <= 192 ||
+    event.keyCode >= 160 && event.keyCode <= 192 ||
     event.keyCode >= 219 && event.keyCode <= 223 ||
     event.keyCode == 252;
   try {
-- 
1.8.1-zedat

