--- a/source3/include/smb_macros.h	2013-02-03 23:52:55.265967765 +0100
+++ b/source3/include/smb_macros.h	2013-02-03 23:54:21.532055283 +0100
@@ -190,7 +190,10 @@
  Check to see if we are a DC for this domain
 *****************************************************************************/
 
-#define IS_DC  (lp_server_role()==ROLE_DOMAIN_PDC || lp_server_role()==ROLE_DOMAIN_BDC || lp_server_role() == ROLE_ACTIVE_DIRECTORY_DC) 
+#define IS_DC  (lp_server_role() == ROLE_DOMAIN_PDC || \
+		lp_server_role() == ROLE_DOMAIN_BDC || \
+		lp_server_role() == ROLE_ACTIVE_DIRECTORY_DC || \
+		lp_server_role() == ROLE_DOMAIN_MEMBER)
 
 /*
  * If you add any entries to KERBEROS_VERIFY defines, please modify the below expressions
