--- a/src/lib/crypto/krb/prng.c
+++ b/src/lib/crypto/krb/prng.c
@@ -138,10 +138,6 @@
        guaranteed to cause a reseed of some kind if strong is true and
        we have both /dev/random and /dev/urandom.  We want the strong
        data included in the reseed so we get it first.*/
-    if (strong) {
-        if (add_entropy_from_device(context, "/dev/random"))
-            *oursuccess = 1;
-    }
     if (add_entropy_from_device(context, "/dev/urandom"))
         *oursuccess = 1;
     return 0;
