Index: samba-3.3.1/source/configure.in
--- configure.in.orig    2009-02-24 00:38:16.000000000 -0700
+++ configure.in    2009-03-11 18:51:10.196201793 -0600
@@ -1223,24 +1223,33 @@
 #
 case "$host_os" in
     *linux*)
-       # glibc <= 2.3.2 has a broken getgrouplist
-       AC_TRY_RUN([
+AC_CACHE_CHECK([for a broken Linux getgrouplist API],
+    linux_getgrouplist_ok,
+    [
+        AC_TRY_RUN([
 #include <unistd.h>
 #include <sys/utsname.h>
-main() {
+
+int main() {
        /* glibc up to 2.3 has a broken getgrouplist */
-#if defined(__GLIBC__) && defined(__GLIBC_MINOR__)
-       int libc_major = __GLIBC__;
-       int libc_minor = __GLIBC_MINOR__;
-
-       if (libc_major < 2)
-              exit(1);
-       if ((libc_major == 2) && (libc_minor <= 3))
-              exit(1);
-#endif
-       exit(0);
-}
-], [linux_getgrouplist_ok=yes], [linux_getgrouplist_ok=no])
+    #if defined(__GLIBC__) && defined(__GLIBC_MINOR__)
+               int libc_major = __GLIBC__;
+               int libc_minor = __GLIBC_MINOR__;
+
+               if (libc_major < 2)
+                      return 1;
+               if ((libc_major == 2) && (libc_minor <= 3))
+                      return 1;
+        #endif
+                   return 0;
+        }
+    ], 
+    
+    [linux_getgrouplist_ok=yes], 
+    [linux_getgrouplist_ok=no]
+    [linux_getgrouplist_ok=cross])
+])
+    
        if test x"$linux_getgrouplist_ok" = x"yes"; then
           AC_DEFINE(HAVE_GETGROUPLIST, 1, [Have good getgrouplist])
        fipatch for ./client/mtab.c
@@ -2530,7 +2530,7 @@
 	# deal, since we can't guarantee that the results we get now will
 	# match the results we get at runtime anyway.
 	if test x"$samba_cv_HAVE_NATIVE_ICONV" = x"cross" ; then
-	    default_dos_charset="CP850"
+	    default_dos_charset="ISO-8859-1"
 	    default_display_charset="ASCII"
 	    default_unix_charset="UTF-8"
 	    samba_cv_HAVE_NATIVE_ICONV=yes
