--- samba-3.3.15/source/lib/sendfile.c.orig	2011-02-27 18:49:27.000000000 +0100
+++ samba-3.3.15/source/lib/sendfile.c		2011-04-28 01:07:35.817609888 +0200
@@ -63,7 +63,11 @@
 #else
 			nwritten = sendfile(tofd, fromfd, &offset, total);
 #endif
-		} while (nwritten == -1 && errno == EINTR);
+#if defined(EWOULDBLOCK)
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK));
+#else
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN));
+#endif
 		if (nwritten == -1) {
 			if (errno == ENOSYS || errno == EINVAL) {
 				/* Ok - we're in a world of pain here. We just sent
@@ -78,8 +82,12 @@
 			}
 			return -1;
 		}
-		if (nwritten == 0)
-			return -1; /* I think we're at EOF here... */
+		if (nwritten == 0) {
+			/*
+			 * EOF, return a short read
+			 */
+			return hdr_len + (count - total);
+		}
 		total -= nwritten;
 	}
 	return count + hdr_len;
@@ -141,7 +149,11 @@
 		int32 nwritten;
 		do {
 			nwritten = sendfile(tofd, fromfd, &small_offset, small_total);
-		} while (nwritten == -1 && errno == EINTR);
+#if defined(EWOULDBLOCK)
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK));
+#else
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN));
+#endif
 		if (nwritten == -1) {
 			if (errno == ENOSYS || errno == EINVAL) {
 				/* Ok - we're in a world of pain here. We just sent
@@ -156,8 +168,12 @@
 			}
 			return -1;
 		}
-		if (nwritten == 0)
-			return -1; /* I think we're at EOF here... */
+		if (nwritten == 0) {
+			/*
+			 * EOF, return a short read
+			 */
+			return hdr_len + (((uint32)count) - small_total);
+		}
 		small_total -= nwritten;
 	}
 	return count + hdr_len;
@@ -218,7 +234,11 @@
 #else
 			nwritten = sendfilev(tofd, vec, sfvcnt, &xferred);
 #endif
-		if (nwritten == -1 && errno == EINTR) {
+#if defined(EWOULDBLOCK)
+		if  (nwritten == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK)) {
+#else
+		if (nwritten == -1 && (errno == EINTR || errno == EAGAIN)) {
+#endif
 			if (xferred == 0)
 				continue; /* Nothing written yet. */
 			else
@@ -266,7 +286,7 @@
 
 	if (header) {
 		/* Set up the header/trailer iovec. */
-		hdtrl[0].iov_base = header->data;
+		hdtrl[0].iov_base = (void *)header->data;
 		hdtrl[0].iov_len = hdr_len = header->length;
 	} else {
 		hdtrl[0].iov_base = NULL;
@@ -292,7 +312,11 @@
 #else
 			nwritten = sendfile(tofd, fromfd, offset, total, &hdtrl[0], 0);
 #endif
-		} while (nwritten == -1 && errno == EINTR);
+#if defined(EWOULDBLOCK)
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK));
+#else
+		} while (nwritten == -1 && (errno == EINTR || errno == EAGAIN));
+#endif
 		if (nwritten == -1)
 			return -1;
 		if (nwritten == 0)
@@ -312,7 +336,7 @@
 				hdtrl[0].iov_len = 0;
 			} else {
 				/* iov_base is defined as a void *... */
-				hdtrl[0].iov_base = ((char *)hdtrl[0].iov_base) + nwritten;
+				hdtrl[0].iov_base = (void *)(((char *)hdtrl[0].iov_base) + nwritten);
 				hdtrl[0].iov_len -= nwritten;
 				nwritten = 0;
 			}
@@ -343,7 +367,7 @@
 
 	/* Set up the header iovec. */
 	if (header) {
-		hdtrl.iov_base = header->data;
+		hdtrl.iov_base = (void *)header->data;
 		hdtrl.iov_len = hdr_len = header->length;
 	} else {
 		hdtrl.iov_base = NULL;
@@ -363,7 +387,11 @@
 
 		do {
 			ret = sendfile(fromfd, tofd, offset, total, &hdr, &nwritten, 0);
-		} while (ret == -1 && errno == EINTR);
+#if defined(EWOULDBLOCK)
+		} while (ret == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK));
+#else
+		} while (ret == -1 && (errno == EINTR || errno == EAGAIN));
+#endif
 		if (ret == -1)
 			return -1;
 
@@ -384,7 +412,7 @@
 				hdtrl.iov_len = 0;
 			} else {
 				hdtrl.iov_base =
-				    (caddr_t)hdtrl.iov_base + nwritten;
+				    (void *)((caddr_t)hdtrl.iov_base + nwritten);
 				hdtrl.iov_len -= nwritten;
 				nwritten = 0;
 			}
@@ -441,7 +469,11 @@
 		*/
 		do {
 			ret = send_file(&tofd, &hdtrl, 0);
-		} while ( (ret == 1) || (ret == -1 && errno == EINTR) );
+#if defined(EWOULDBLOCK)
+		} while ((ret == 1) || (ret == -1 && (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK)));
+#else
+		} while ((ret == 1) || (ret == -1 && (errno == EINTR || errno == EAGAIN)));
+#endif
 		if ( ret == -1 )
 			return -1;
 	}
