--- a/src/plugins/kdb/ldap/libkdb_ldap/ldap_krbcontainer.c	2014-08-05 13:46:28.984116945 -0500
+++ b/src/plugins/kdb/ldap/libkdb_ldap/ldap_krbcontainer.c	2014-08-05 13:57:24.048490846 -0500
@@ -32,6 +32,8 @@
 #include "kdb_ldap.h"
 #include "ldap_err.h"
 
+char    *policyrefattribute[] = {"krbTicketPolicyReference",NULL};
+
 /*
  * Read the kerberos container location from krb5.conf.
  */
@@ -41,8 +41,10 @@
 krb5_error_code
 krb5_ldap_read_krbcontainer_dn(krb5_context context, char **container_dn)
 {
-    krb5_error_code                 st=0;
+    krb5_error_code                 st=0, tempst=0;
     char                            *dn=NULL;
+    LDAP                            *ld=NULL;
+    LDAPMessage                     *result=NULL, *ent=NULL;
     kdb5_dal_handle                 *dal_handle=NULL;
     krb5_ldap_context               *ldap_context=NULL;
     krb5_ldap_server_handle         *ldap_server_handle=NULL;
@@ -74,6 +77,25 @@
         }
     }
 
+    /* try to locate container in LDAP */
+    if (dn == NULL) {
+        LDAP_SEARCH_1(NULL, LDAP_SCOPE_SUBTREE, "(objectclass=krbContainer)", policyrefattribute, IGNORE_STATUS);
+        if (st != LDAP_SUCCESS && st != LDAP_NO_SUCH_OBJECT) {
+            st = set_ldap_error(context, st, OP_SEARCH);
+            goto cleanup;
+        }
+
+        if (st == LDAP_NO_SUCH_OBJECT) {
+            st = KRB5_KDB_NOENTRY;
+            goto cleanup;
+        }
+
+        if ((ent = ldap_first_entry(ld, result))) {
+            dn = ldap_get_dn(ld, ent);
+            ldap_msgfree(result);
+        }
+    }
+
     if (dn == NULL) {
         st = KRB5_KDB_SERVER_INTERNAL_ERR;
         krb5_set_error_message(context, st,
