--- a/src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c
+++ b/src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c
@@ -1356,8 +1356,8 @@
         goto cleanup;
 
     for (i=0; bvalues[i] != NULL; ++i) {
-        krb5_int16 n_kd;
-        krb5_key_data *kd;
+        krb5_int16 n_kd=0;
+        krb5_key_data *kd=NULL;
         krb5_data in;
 
         if (bvalues[i]->bv_len == 0)
--- a/src/kdc/do_as_req.c
+++ b/src/kdc/do_as_req.c
@@ -135,7 +135,7 @@
     const char *emsg = 0;
     int did_log = 0;
     register int i;
-    krb5_enctype useenctype;
+    krb5_enctype useenctype = 0;
     loop_respond_fn oldrespond;
     void *oldarg;
     kdc_realm_t *kdc_active_realm = state->active_realm;
--- a/src/slave/kprop.c
+++ b/src/slave/kprop.c
@@ -92,7 +92,7 @@
     int     argc;
     char    **argv;
 {
-    int     fd, database_fd, database_size;
+    int     fd=0, database_fd, database_size;
     krb5_error_code retval;
     krb5_context context;
     krb5_creds *my_creds;
