--- a/cmulocal/sasl2.m4
+++ b/cmulocal/sasl2.m4
@@ -66,7 +66,7 @@
     fi
   fi
   AC_CHECK_HEADER([gssapi.h],,
-                  [AC_CHECK_HEADER([gssapi/gssapi.h],,
+                  [AC_CHECK_HEADER([krb5/gssapi/gssapi.h],,
                                    [AC_WARN([Disabling GSSAPI - no include files found]); gssapi=no])])
   AC_CHECK_HEADERS(gssapi/gssapi_ext.h)
   CPPFLAGS=$cmu_saved_CPPFLAGS
