--- a/src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c	2012-07-11 20:05:00.225834136 +0200
+++ b/src/plugins/kdb/ldap/libkdb_ldap/ldap_principal2.c	2012-07-11 20:05:39.134538384 +0200
@@ -1356,8 +1356,8 @@
         goto cleanup;
 
     for (i=0; bvalues[i] != NULL; ++i) {
-        krb5_int16 n_kd;
-        krb5_key_data *kd;
+        krb5_int16 n_kd=0;
+        krb5_key_data *kd=NULL;
         krb5_data in;
 
         if (bvalues[i]->bv_len == 0)
--- a/src/kdc/do_as_req.c	2012-07-11 20:10:21.531380063 +0200
+++ b/src/kdc/do_as_req.c	2012-07-11 20:08:17.670327693 +0200
@@ -135,7 +135,7 @@
     const char *emsg = 0;
     int did_log = 0;
     register int i;
-    krb5_enctype useenctype;
+    krb5_enctype useenctype = 0;
     loop_respond_fn oldrespond;
     void *oldarg;
     kdc_realm_t *kdc_active_realm = state->active_realm;
--- a/src/slave/kprop.c	2012-07-11 20:11:26.897878314 +0200
+++ b/src/slave/kprop.c	2012-07-11 20:12:42.204298183 +0200
@@ -90,7 +90,7 @@
 int
 main(int argc, char **argv)
 {
-    int fd, database_fd, database_size;
+    int fd=0, database_fd, database_size;
     krb5_error_code retval;
     krb5_context context;
     krb5_creds *my_creds;
--- a/src/lib/krb5/os/sendto_kdc.c
+++ b/src/lib/krb5/os/sendto_kdc.c
@@ -339,7 +339,7 @@
     struct timeval tv;
 #endif
     krb5_error_code retval;
-    time_ms curtime, interval;
+    time_ms curtime=0, interval;
 
     retval = get_curtime_ms(&curtime);
     if (retval != 0)
@@ -1315,7 +1315,7 @@
             void *msg_handler_data, struct conn_state **winner_out)
 {
     int e, selret = 0;
-    time_ms endtime;
+    time_ms endtime=0;
     struct conn_state *state;
 
     *winner_out = NULL;
