--- a/src/plugins/kdb/ldap/libkdb_ldap/kdb_ldap_conn.c	2014-08-05 12:41:24.255016325 -0500
+++ b/src/plugins/kdb/ldap/libkdb_ldap/kdb_ldap_conn.c	2014-08-05 12:53:41.319523332 -0500
@@ -44,13 +44,13 @@
     krb5_error_code             st=0;
     unsigned char               *password=NULL;
 
-    if (ldap_context->bind_dn == NULL) {
+    if (ldap_context->bind_dn == NULL && ldap_context->bind_pwd != NULL) {
         st = EINVAL;
         krb5_set_error_message(context, st, _("LDAP bind dn value missing "));
         goto err_out;
     }
 
-    if (ldap_context->bind_pwd == NULL && ldap_context->service_password_file == NULL) {
+    if (ldap_context->bind_dn != NULL && ldap_context->bind_pwd == NULL && ldap_context->service_password_file == NULL) {
         st = EINVAL;
         krb5_set_error_message(context, st,
                                _("LDAP bind password value missing "));
@@ -90,6 +90,34 @@
 {
     struct berval               bv={0, NULL};
 
+    if (ldap_context->bind_dn == NULL) {
+        /* Certificate based or UID/GID based bind (SASL EXTERNAL mechanism) */
+
+        krb5_error_code st;
+        struct berval      *servercreds=NULL;
+
+        st = ldap_sasl_bind_s(ldap_server_handle->ldap_handle,
+                              NULL,        /* Authenticating dn */
+                              "EXTERNAL",  /* Method used for authentication */
+                              &bv,
+                              NULL,
+                              NULL,
+                              &servercreds);
+
+        if (st == LDAP_SASL_BIND_IN_PROGRESS) {
+            st = ldap_sasl_bind_s(ldap_server_handle->ldap_handle,
+                                  NULL,
+                                  "EXTERNAL",
+                                  servercreds,
+                                  NULL,
+                                  NULL,
+                                  &servercreds);
+        }
+
+        return st;
+    } 
+
+    /* password based simple bind */
     bv.bv_val = ldap_context->bind_pwd;
     bv.bv_len = strlen(ldap_context->bind_pwd);
     return ldap_sasl_bind_s(ldap_server_handle->ldap_handle,
